File: Cpp/src/parser.cpp
//
// Created by irwin on 12/05/2021.
//
#include "../header/parser.h"

namespace corny {
    void Parser::advance(TokenType type) {
        if (curToken.type == type){
            nextToken();
            return;
        }
        std::cout << "Lexer Error: Unexpected token. Got: " << curToken.type << ", want: " << type << std::endl;
    }
    void Parser::nextToken() {
        curToken = peekToken;
        peekToken = lexer.nextToken();
    }
    // program ::= ( statement )*
    Node* Parser::parseProgram() {
        ProgramNode* program = new ProgramNode();

        while (curToken.type != TT_EOF) {
            program->statements.emplace_back(parseStatement());
        }

        return program;
    }
    // parseBlock ::= '{' (statement)* '}'
    Node* Parser::parseBlock() {
        BlockNode* blockNode = new BlockNode();

        advance(TT_LBRACE);
        while (curToken.type != TT_RBRACE) {
            blockNode->statements.emplace_back(parseStatement());
        }
        advance(TT_RBRACE);

        return blockNode;
    }
    // parseStatement ::= parseLetStatement | parseReturnStatement | parseExpression
    Node* Parser::parseStatement() {
        Node* statement;
        if (curToken.type == TT_LET) {
            statement = parseLetStatement();
        }
        else if (curToken.type == TT_RETURN) {
            statement = parseReturnStatement();
        }
        else {
            statement = parseExpression();
        }
        // optionally skip the semicolon
        if (curToken.type == TT_SEMICOLON) {
            advance(TT_SEMICOLON);
        }
        return statement;
    }
    // parseLetStatement ::= 'let' IDENT '=' parseExpression
    Node* Parser::parseLetStatement() {
        LetNode* letNode = new LetNode();

        advance(TT_LET);

        letNode->ident = (IdentNode*)parseIdentifier();
        advance(TT_ASSIGN);
        letNode->value = parseExpression();

        return letNode;
    }
    // parseReturnStatement ::= 'return' parseExpression
    Node* Parser::parseReturnStatement() {
        ReturnNode* returnNode = new ReturnNode();

        advance(TT_RETURN);
        returnNode->value = parseExpression();

        return returnNode;
    }
    // parseExpression ::= parseAssignment
    Node* Parser::parseExpression() {
        return parseAssignment();
    }
    // parseAssignment ::= parseLogicOr
    Node* Parser::parseAssignment() {
        return parseLogicOr();
    }
    // parseLogicOr ::= parseLogicAnd ('and' parseLogicAnd)*
    Node* Parser::parseLogicOr() {
        Node* node = parseLogicAnd();
        while (curToken.type == TT_OR) {
            Token token = curToken;
            advance(TT_OR);
            node = new BinOpNode(node, token, parseLogicAnd());
        }
        return node;
    }
    // parseLogicAnd ::= parseEquality ('and' parseEquality)*
    Node* Parser::parseLogicAnd() {
        Node* node = parseEquality();
        while (curToken.type == TT_AND) {
            Token token = curToken;
            advance(TT_AND);
            node = new BinOpNode(node, token, parseEquality());
        }
        return node;
    }
    // parseEquality ::= parseComparison (('==', '!=') parseComparison)*
    Node* Parser::parseEquality() {
        Node* node = parseComparison();
        while (curToken.type == TT_EQUAL || curToken.type == TT_NOT_EQ) {
            Token token = curToken;
            advance(token.type);
            node = new BinOpNode(node, token, parseComparison());
        }
        return node;
    }
    // parseComparison ::= parseTerm (('<'|'<='|'>'|'>=') parseTerm)*
    Node* Parser::parseComparison() {
        Node* node = parseTerm();
        while (curToken.type == TT_LESS || curToken.type == TT_LESS_EQ ||
        curToken.type == TT_GREATER || curToken.type == TT_GREATER_EQ)
        {
            Token token = curToken;
            advance(token.type);
            node = new BinOpNode(node, token, parseTerm());
        }
        return node;
    }
    // parseTerm ::= parseFactor (('+'|'-') parseFactor)*
    Node* Parser::parseTerm() {
        Node* node = parseFactor();
        while (curToken.type == TT_PLUS || curToken.type == TT_MINUS) {
            Token token = curToken;
            advance(token.type);
            node = new BinOpNode(node, token, parseFactor());
        }
        return node;
    }
    // parseFactor ::= parseUnary (('*'|'/' parseUnary)*
    Node* Parser::parseFactor() {
        Node* node = parseUnary();
        while (curToken.type == TT_MUL || curToken.type == TT_DIV) {
            Token token = curToken;
            advance(token.type);
            node = new BinOpNode(node, token, parseUnary());
        }
        return node;
    }
    // parseUnary ::= ('-'|'!') parseUnary | parseCall
    Node* Parser::parseUnary() {
        while (curToken.type == TT_MINUS || curToken.type == TT_NOT) {
            Token token = curToken;
            advance(token.type);
            return new UnaryNode(token, parseUnary()); // call itself
        }
        return parseCall();
    }
    // parseCall ::= parsePrimary | parseCallExpression
    Node* Parser::parseCall() {
        Node* node = parsePrimary();
        while (true) {
            if (curToken.type == TT_LPAREN || curToken.type == TT_LBRACKET) {
                node = parseCallExpr(node, curToken);
            }
            else if (curToken.type == TT_QUESTION) {
                node = parseTernaryExpr(node);
            }
            else {
                break;
            }
        }
        return node;
    }
    // parsePrimary ::= NUMBER | TRUE | FALSE | STRING | NULL | IDENT | FUNCTION | IF-ELSE
    Node* Parser::parsePrimary() {
        Token token = curToken;
        switch (token.type) {
            case TT_NUMBER:
                advance(TT_NUMBER);
                return new NumberNode(std::stod(token.literal));
            case TT_TRUE:
                advance(TT_TRUE);
                return new BooleanNode(true);
            case TT_FALSE:
                advance(TT_FALSE);
                return new BooleanNode(false);
            case TT_NULL:
                advance(TT_NULL);
                return new NullNode();
            case TT_IDENT:
                return parseIdentifier();
            case TT_LPAREN: {
                advance(TT_LPAREN);
                Node* resultExpr = parseExpression();
                advance(TT_RPAREN);
                return resultExpr;
            }
            case TT_STRING:
                advance(TT_STRING);
                return new StringNode(token.literal);
            case TT_FUNCTION:
                return parseFunctionLiteral();
            case TT_LBRACKET:
                return parseArrayLiteral();
            case TT_LBRACE:
                return parseHashLiteral();
            case TT_IF:
                return parseIfExpr();
            default:
                std::cout << "Unknown expression token: " << token.literal << std::endl;
                std::exit(1);
        }
    }
    // parseCallExpr ::= FUNCTIONCALL | ARRAYCALL | HASHCALL
    Node* Parser::parseCallExpr(Node *callee, Token token) {
        CallExprNode *callExprNode = new CallExprNode();
        callExprNode->callee = callee;

        if (token.type == TT_LBRACKET) { // array or hash call
            advance(TT_LBRACKET);
            if (curToken.type == TT_RBRACKET) {
                std::cout << "Invalid subscript reference for array or hash types\n";
                std::exit(1);
            }
            callExprNode->arguments.emplace_back(parseExpression());
            advance(TT_RBRACKET);
        }
        else if (token.type == TT_LPAREN) { // function call
            advance(TT_LPAREN);
            if (curToken.type != TT_RPAREN) {
                callExprNode->arguments.emplace_back(parseExpression());
                while (curToken.type == TT_COMMA) {
                    advance(TT_COMMA);
                    callExprNode->arguments.emplace_back(parseExpression());
                }
            }
            advance(TT_RPAREN);
        }
        return callExprNode;
    }
    // parseTernaryExpr
    Node* Parser::parseTernaryExpr(Node *condition) {
        IfNode *ifNode = new IfNode();

        ifNode->condition = parseExpression();

        advance(TT_QUESTION); // advance '?'
        ifNode->consequence = parseBlock();

        advance(TT_COLON);
        ifNode->alternative = parseBlock();

        return ifNode;
    }
    // parseFunctionLiteral
    Node* Parser::parseFunctionLiteral() {
        FunctionNode *functionNode = new FunctionNode();
        advance(TT_FUNCTION);
        advance(TT_LPAREN);
        if (curToken.type != TT_RPAREN) {
            functionNode->parameters.emplace_back((IdentNode*)parseIdentifier());
            while (curToken.type == TT_COMMA) {
                advance(TT_COMMA);
                functionNode->parameters.emplace_back((IdentNode*)parseIdentifier());
            }
        }
        advance(TT_RPAREN);
        functionNode->body = (BlockNode*)parseBlock();

        return functionNode;
    }
    // parseArrayLiteral
    Node* Parser::parseArrayLiteral() {
        ArrayNode* arrayNode = new ArrayNode();
        advance(TT_LBRACKET);
        if (curToken.type != TT_RBRACKET) {
            arrayNode->elements.emplace_back(parseExpression());
            while (curToken.type == TT_COMMA) {
                advance(TT_COMMA);
                arrayNode->elements.emplace_back(parseExpression());
            }
        }
        advance(TT_RBRACKET);

        return arrayNode;
    }
    // parseHashLiteral
    Node* Parser::parseHashLiteral() {
        HashNode* hashNode = new HashNode();
        advance(TT_LBRACE);
        if (curToken.type != TT_RBRACE) {
            // parse key-value pair expressions
            hashNode->keys.emplace_back((IdentNode*)parseExpression());
            advance(TT_COLON); // advance ':'
            hashNode->values.emplace_back(parseExpression());

            while (curToken.type == TT_COMMA) {
                advance(TT_COMMA);
                // parse key-value pair expressions (refactor here in the future)
                hashNode->keys.emplace_back((IdentNode*)parseExpression());
                advance(TT_COLON); // advance ':'
                hashNode->values.emplace_back(parseExpression());
            }
        }
        advance(TT_RBRACE);
        return hashNode;
    }
    // parseIfExpression
    Node* Parser::parseIfExpr() {
        IfNode* ifNode = new IfNode();

        advance(TT_IF);

        advance(TT_LPAREN);
        ifNode->condition = parseExpression();
        advance(TT_RPAREN);

        ifNode->consequence = parseBlock();

        if (curToken.type = TT_ELSE) {
            advance(TT_ELSE);
            ifNode->alternative = parseBlock();
        }

        return ifNode;
    }
    // parserIdentifier
    Node* Parser::parseIdentifier() {
        IdentNode* identNode = new IdentNode();
        identNode->value = curToken;

        advance(TT_IDENT); // skip the IDENT token.

        // new feature will be placed here...

        return identNode;
    }
}

File: Go/parser/parser.go
package parser

import (
	"CornyLang/ast"
	"CornyLang/lexer"
	"CornyLang/token"
	"fmt"
	"os"
	"strconv"
)

type Parser struct {
	lexer     *lexer.Lexer
	curToken  token.Token
	peekToken token.Token
	errors    []string
}

func New(lexer *lexer.Lexer) *Parser {
	parser := &Parser{lexer: lexer}
	parser.nextToken() // feed peekToken
	parser.nextToken() // feed curToken
	return parser
}

func (p *Parser) nextToken() {
	p.curToken = p.peekToken
	p.peekToken = p.lexer.NextToken()
}

func (p *Parser) advance(tokenType token.TokenType) {
	if p.curToken.Type == tokenType {
		p.nextToken()
	} else {
		msg := fmt.Sprintf("unexpected token. got %s instead of %s", tokenType, p.curToken.Type)
		p.errors = append(p.errors, msg)
	}
}

// program ::= ( statement )*
func (p *Parser) Program() *ast.ProgramNode {
	var programNode = &ast.ProgramNode{}
	programNode.Statements = []ast.Statement{}

	for p.curToken.Type != token.TT_EOF {
		var stmt = p.parseStatement()
		if stmt != nil {
			programNode.Statements = append(programNode.Statements, stmt)
		}
	}

	return programNode
}

// parseBlockStmt ::= LBRACE ( statement )* RBRACE
func (p *Parser) parseBlockStmt() *ast.BlockStmtNode {
	var blockNode = &ast.BlockStmtNode{Token: p.curToken}

	blockNode.Statements = []ast.Statement{}

	p.advance(token.TT_LBRACE)
	for p.curToken.Type != token.TT_RBRACE {
		blockNode.Statements = append(blockNode.Statements, p.parseStatement())
	}
	p.advance(token.TT_RBRACE)

	return blockNode
}

// statement ::= parseLetStatement
// 			  	| parseReturnStatement
// 			  	| parseFunctionStatement
//				| parseExpressionStatement
func (p *Parser) parseStatement() ast.Statement {
	switch p.curToken.Type {
	case token.TT_LET:
		return p.parseLetStmt()
	case token.TT_RETURN:
		return p.parseReturnStmt()
	default:
		return p.parseExpressionStmt()
	}

}

// parseLetStmt ::= let IDENTIFIER = parseExpression
func (p *Parser) parseLetStmt() ast.Statement {
	var letNode = &ast.LetNode{Token: p.curToken}

	p.advance(token.TT_LET)
	letNode.Name = p.parseIdentifier()

	p.advance(token.TT_ASSIGN)
	letNode.Value = p.parseExpression()

	if p.curToken.Type == token.TT_SEMICOLON {
		p.advance(token.TT_SEMICOLON)
	}

	return letNode
}

// parseReturnStmt
func (p *Parser) parseReturnStmt() ast.Statement {
	var returnNode = &ast.ReturnNode{Token: p.curToken}
	p.advance(token.TT_RETURN)

	returnNode.Value = p.parseExpression()

	if p.curToken.Type == token.TT_SEMICOLON {
		p.advance(token.TT_SEMICOLON)
	}

	return returnNode
}

// parseExpressionStmt ::= logicOr
func (p *Parser) parseExpressionStmt() *ast.ExpressionStatement {
	var expressionStmt = &ast.ExpressionStatement{}
	expressionStmt.Expression = p.parseExpression()
	if p.curToken.Type == token.TT_SEMICOLON {
		p.advance(token.TT_SEMICOLON)
	}
	return expressionStmt
}

func (p *Parser) parseExpression() ast.Expression {
	return p.logicOr()
}

func (p *Parser) logicOr() ast.Expression {
	var node = p.logicAnd()
	for p.curToken.Type == token.TT_OR {
		var token = p.curToken
		p.advance(token.Type)
		node = &ast.BinOpNode{Left: node, Operator: token, Right: p.logicAnd()}
	}
	return node
}

func (p *Parser) logicAnd() ast.Expression {
	var node = p.equality()
	for p.curToken.Type == token.TT_AND {
		var token = p.curToken
		p.advance(token.Type)
		node = &ast.BinOpNode{Left: node, Operator: token, Right: p.equality()}
	}
	return node
}

func (p *Parser) equality() ast.Expression {
	var node = p.comparison()
	for p.curToken.Type == token.TT_EQUAL || p.curToken.Type == token.TT_NOT_EQ {
		var token = p.curToken
		p.advance(token.Type)
		node = &ast.BinOpNode{Left: node, Operator: token, Right: p.comparison()}
	}
	return node
}

func (p *Parser) comparison() ast.Expression {
	var node = p.term()
	for p.curToken.Type == token.TT_LESS ||
		p.curToken.Type == token.TT_LESS_EQ ||
		p.curToken.Type == token.TT_GREATER ||
		p.curToken.Type == token.TT_GREATER_EQ {
		var token = p.curToken
		p.advance(token.Type)
		node = &ast.BinOpNode{Left: node, Operator: token, Right: p.term()}
	}
	return node
}

func (p *Parser) term() ast.Expression {
	var node = p.factor()
	for p.curToken.Type == token.TT_PLUS || p.curToken.Type == token.TT_MINUS {
		var token = p.curToken
		p.advance(token.Type)
		node = &ast.BinOpNode{Left: node, Operator: token, Right: p.factor()}
	}
	return node
}

func (p *Parser) factor() ast.Expression {
	var node = p.unary()
	for p.curToken.Type == token.TT_MUL || p.curToken.Type == token.TT_DIV {
		var token = p.curToken
		p.advance(token.Type)
		node = &ast.BinOpNode{Left: node, Operator: token, Right: p.unary()}
	}
	return node
}

// unary ::= ( ( '-' | '!' ) unary )* | call
func (p *Parser) unary() ast.Expression {
	for p.curToken.Type == token.TT_MINUS || p.curToken.Type == token.TT_NOT {
		var token = p.curToken
		p.advance(token.Type)
		return &ast.UnaryNode{Operator: token, Right: p.unary()}
	}
	return p.call()
}

// call ::= primary '(' arguments? ')'
func (p *Parser) call() ast.Expression {
	var node = p.primary()
	for {
		if p.curToken.Type == token.TT_LPAREN ||
			p.curToken.Type == token.TT_LBRACKET ||
			p.curToken.Type == token.TT_DOT {
			node = p.parseCallExpr(node)
		} else if p.curToken.Type == token.TT_QUESTION {
			node = p.parseTernaryExpr(node)
		} else {
			break
		}
	}
	return node
}

// primary ::= NUMBRE | STRING | TRUE | FALSE | NULL | IDENT | FUNCTION | IF-ELSE
func (p *Parser) primary() ast.Expression {
	var tok = p.curToken
	switch tok.Type {
	case token.TT_NUMBER:
		p.advance(token.TT_NUMBER)
		value, _ := strconv.ParseFloat(tok.Literal, 64)
		return &ast.NumberNode{Token: tok, Value: value}
	case token.TT_TRUE:
		p.advance(token.TT_TRUE)
		return &ast.BooleanNode{Token: tok, Value: true}
	case token.TT_FALSE:
		p.advance(token.TT_FALSE)
		return &ast.BooleanNode{Token: tok, Value: false}
	case token.TT_NULL:
		p.advance(token.TT_NULL)
		return &ast.NullNode{Token: tok}
	case token.TT_IDENT:
		return p.parseIdentifier()
	case token.TT_LPAREN:
		p.advance(token.TT_LPAREN)
		var resultExpr = p.parseExpression()
		p.advance(token.TT_RPAREN)
		return resultExpr
	case token.TT_STRING:
		p.advance(token.TT_STRING)
		return &ast.StringNode{Token: tok, Value: tok.Literal}
	case token.TT_FUNCTION:
		return p.parseFunctionLiteral()
	case token.TT_LBRACKET:
		return p.parseArrayLiteral()
	case token.TT_LBRACE:
		return p.parseHashLiteral()
	case token.TT_IF:
		return p.parseIfExpr()
	default:
		fmt.Printf("Unknown expression token: %s\n", tok.Literal)
		os.Exit(4)
	}
	return nil
}

// parseCallExpr ::= FUNCTIONCALL | ARRAYCALL | HASHCALL
func (p *Parser) parseCallExpr(callee ast.Expression) ast.Expression {
	var callExprNode = &ast.CallExprNode{Token: p.curToken}
	callExprNode.Callee = callee
	callExprNode.Arguments = []ast.Expression{}

	if p.curToken.Type == token.TT_LBRACKET { // array or hash call
		p.advance(token.TT_LBRACKET)
		if p.curToken.Type == token.TT_RBRACKET {
			fmt.Print("Invalid subscript reference for array or hash types\n")
			os.Exit(4)
		}

		callExprNode.Arguments = append(callExprNode.Arguments, p.parseExpression())
		p.advance(token.TT_RBRACKET)
	} else if p.curToken.Type == token.TT_LPAREN { // function call
		p.advance(token.TT_LPAREN)
		if p.curToken.Type != token.TT_RPAREN {
			callExprNode.Arguments = append(callExprNode.Arguments, p.parseExpression())
			for p.curToken.Type == token.TT_COMMA {
				p.advance(token.TT_COMMA)
				callExprNode.Arguments = append(callExprNode.Arguments, p.parseExpression())
			}
		}
		p.advance(token.TT_RPAREN)
	} else if p.curToken.Type == token.TT_DOT {
		p.advance(token.TT_DOT)
		callExprNode.Arguments = append(callExprNode.Arguments, p.parseExpression())
	}
	return callExprNode
}

// parseTernaryExpr
func (p *Parser) parseTernaryExpr(conditionNode ast.Expression) ast.Expression {
	var ifNode = &ast.IfExprNode{Token: p.curToken}
	p.advance(token.TT_QUESTION)

	ifNode.Condition = conditionNode
	ifNode.Consequence = &ast.BlockStmtNode{Statements: []ast.Statement{p.parseStatement()}}

	p.advance(token.TT_COLON)
	ifNode.Alternative = &ast.BlockStmtNode{Statements: []ast.Statement{p.parseStatement()}}

	return ifNode
}

// parseFunctionLiteral
func (p *Parser) parseFunctionLiteral() ast.Expression {
	var functionNode = &ast.FunctionLiteralNode{Token: p.curToken}
	p.advance(token.TT_FUNCTION)
	p.advance(token.TT_LPAREN)

	if p.curToken.Type != token.TT_RPAREN {
		functionNode.Parameters = []*ast.IdentifierNode{}
		functionNode.Parameters = append(functionNode.Parameters, p.parseIdentifier())
		for p.curToken.Type == token.TT_COMMA {
			p.advance(token.TT_COMMA)
			functionNode.Parameters = append(functionNode.Parameters, p.parseIdentifier())
		}
	}
	p.advance(token.TT_RPAREN)
	functionNode.Body = p.parseBlockStmt()

	return functionNode
}

// parseIdentifier
func (p *Parser) parseIdentifier() *ast.IdentifierNode {
	var identifierNode = &ast.IdentifierNode{Token: p.curToken}
	identifierNode.Value = p.curToken.Literal

	p.advance(token.TT_IDENT)

	return identifierNode
}

// parseArrayLiteral
func (p *Parser) parseArrayLiteral() *ast.ArrayLiteralNode {
	var arrayNode = &ast.ArrayLiteralNode{Token: p.curToken}

	p.advance(token.TT_LBRACKET)
	if p.curToken.Type != token.TT_RBRACKET {
		arrayNode.Elements = []ast.Expression{}
		arrayNode.Elements = append(arrayNode.Elements, p.parseExpression())
		for p.curToken.Type == token.TT_COMMA {
			p.advance(token.TT_COMMA)
			arrayNode.Elements = append(arrayNode.Elements, p.parseExpression())
		}
	}
	p.advance(token.TT_RBRACKET)

	return arrayNode
}

// parseHashLiteral
func (p *Parser) parseHashLiteral() *ast.HashLiteralNode {
	var hashNode = &ast.HashLiteralNode{Token: p.curToken}
	p.advance(token.TT_LBRACE)
	if p.curToken.Type != token.TT_RBRACE {
		hashNode.Keys = []ast.Expression{}
		hashNode.Values = []ast.Expression{}

		// parse key-value pairs expressions
		hashNode.Keys = append(hashNode.Keys, p.parseExpression())
		p.advance(token.TT_COLON)
		hashNode.Values = append(hashNode.Values, p.parseExpression())

		for p.curToken.Type == token.TT_COMMA {
			p.advance(token.TT_COMMA)
			// parse key-value pairs expressions
			// TODO: apply refactoring in duplicated code.
			hashNode.Keys = append(hashNode.Keys, p.parseExpression())
			p.advance(token.TT_COLON)
			hashNode.Values = append(hashNode.Values, p.parseExpression())
		}

	}
	p.advance(token.TT_RBRACE)
	return hashNode
}

// parseIfExpr
func (p *Parser) parseIfExpr() *ast.IfExprNode {
	var ifNode = &ast.IfExprNode{Token: p.curToken}

	p.advance(token.TT_IF)

	p.advance(token.TT_LPAREN)
	ifNode.Condition = p.parseExpression()
	p.advance(token.TT_RPAREN)

	ifNode.Consequence = p.parseBlockStmt()

	if p.curToken.Type == token.TT_ELSE {
		p.advance(token.TT_ELSE)
		ifNode.Alternative = p.parseBlockStmt()
	}

	return ifNode
}
